File: src/cpp/bindings/arrays.cpp
#include "bindings/registry.hpp"

void register_pfm(nanobind::module_ &);
void register_npy_npz(nanobind::module_ &);
void register_flo(nanobind::module_ &);
void register_safetensors(nanobind::module_ &);
void register_dmb(nanobind::module_ &);

namespace sio::bindings {
namespace {

constexpr RegistrationDescriptor REGISTRATIONS[] = {
    {0, "pfm", &::register_pfm},
    {15, "npy_npz", &::register_npy_npz},
    {19, "flo", &::register_flo},
    {34, "safetensors", &::register_safetensors},
    {35, "dmb", &::register_dmb},
};

constexpr CodecDescriptor CODECS[] = {
    {0,
     "pfm",
     "arrays",
     symbols("read_pfm"),
     symbols("write_pfm"),
     symbols("_inspect_pfm"),
     symbols("read_pfm"),
     symbols("write_pfm"),
     symbols("read_pfm_window")},
    {25,
     "npy",
     "arrays",
     symbols("read_npy", "read_npy_view"),
     symbols("write_npy"),
     symbols("_inspect_npy"),
     symbols("read_npy", "read_npy_view"),
     symbols("write_npy"),
     symbols()},
    {26,
     "npz",
     "arrays",
     symbols("read_npz"),
     symbols("write_npz"),
     symbols(),
     symbols("read_npz"),
     symbols("write_npz"),
     symbols()},
    {27,
     "safetensors",
     "arrays",
     symbols("read_safetensors", "read_safetensors_view"),
     symbols("write_safetensors"),
     symbols("_inspect_safetensors"),
     symbols("read_safetensors", "read_safetensors_view"),
     symbols("write_safetensors"),
     symbols(
         "read_safetensors_tensors",
         "read_safetensors_tensors_view",
         "read_safetensors_slices",
         "read_safetensors_slices_view")},
    {44,
     "flo",
     "arrays",
     symbols("read_flo", "read_flo_view"),
     symbols("write_flo"),
     symbols(),
     symbols("read_flo", "read_flo_view"),
     symbols("write_flo"),
     symbols("read_flo", "read_flo_view")},
    {45,
     "dmb",
     "arrays",
     symbols("read_dmb"),
     symbols("write_dmb"),
     symbols("_inspect_dmb"),
     symbols("read_dmb"),
     symbols("write_dmb"),
     symbols("read_dmb_window")},
};

constexpr FamilyBindings FAMILY{
    "arrays",
    REGISTRATIONS,
    sizeof(REGISTRATIONS) / sizeof(REGISTRATIONS[0]),
    CODECS,
    sizeof(CODECS) / sizeof(CODECS[0]),
};

} // namespace

const FamilyBindings &array_bindings() { return FAMILY; }

} // namespace sio::bindings

File: src/sceneio/io/_registry/families/arrays.py
"""Built-in array, tensor, depth, and flow codec definitions."""

from __future__ import annotations

from sceneio import _core
from sceneio.io._registry.adapters import (
    _array_window_reader,
    _file_sink_writer,
    _mmap_reader,
    _mmap_selector_reader,
    _mmap_view_reader,
)
from sceneio.io._registry.model import Codec


def build_array_codecs(
    _canon,
    _prepare_tensor_dict,
) -> tuple[Codec, ...]:
    """Build the array family with facade-owned preparation callbacks."""

    return (
        Codec(
            "pfm",
            (".pfm",),
            _mmap_reader(_core.read_pfm),
            _file_sink_writer(_core.write_pfm, prepare=_canon),
            record=None,
            datatype="depth_map",
            magic=(b"PF", b"Pf"),
            read_window=_mmap_selector_reader(_core.read_pfm_window),
            supported_features=(
                "grayscale",
                "rgb",
                "float32",
                "little_endian",
                "big_endian",
                "typed_depth_adapter",
            ),
            unsupported_features=("native_positive_stride_mmap_view",),
        ),
        Codec(
            "npy",
            (".npy",),
            _mmap_view_reader(_core.read_npy_view, _core.read_npy),
            _file_sink_writer(_core.write_npy, prepare=_canon),
            record=None,
            datatype="tensor",
            magic=(b"\x93NUMPY",),
            supported_features=("v1", "c_order", "native_endian_mmap_view"),
            unsupported_features=("fortran_order", "object_dtype"),
        ),
        Codec(
            "npz",
            (".npz",),
            _mmap_reader(_core.read_npz),
            _file_sink_writer(_core.write_npz, prepare=_prepare_tensor_dict),
            record=_core.TensorDict,
            datatype="tensor_dict",
            supported_features=("stored", "deflate", "numeric_dtypes"),
            unsupported_features=("object_dtype",),
        ),
        Codec(
            "safetensors",
            (".safetensors",),
            _mmap_view_reader(
                _core.read_safetensors_view,
                _core.read_safetensors,
            ),
            _file_sink_writer(
                _core.write_safetensors,
                prepare=_prepare_tensor_dict,
            ),
            record=_core.TensorDict,
            datatype="tensor_dict",
            read_tensors=_mmap_view_reader(
                _core.read_safetensors_tensors_view,
                _core.read_safetensors_tensors,
            ),
            read_slices=_mmap_view_reader(
                _core.read_safetensors_slices_view,
                _core.read_safetensors_slices,
            ),
            supported_features=(
                "metadata",
                "bool",
                "signed_integers",
                "unsigned_integers",
                "float16",
                "float32",
                "float64",
                "mmap_views",
                "leading_axis_slices",
            ),
            unsupported_features=(
                "bfloat16",
                "float8",
                "complex64",
                "sub_byte_dtypes",
                "strided_tensors",
            ),
        ),
        Codec(
            "flo",
            (".flo",),
            _mmap_view_reader(_core.read_flo_view, _core.read_flo),
            _file_sink_writer(_core.write_flo, prepare=_canon),
            record=None,
            datatype="flow",
            magic=(b"PIEH",),
            read_window=_array_window_reader(
                _mmap_view_reader(_core.read_flo_view, _core.read_flo)
            ),
            supported_features=(
                "float32",
                "native_endian_mmap_view",
                "typed_flow_adapter",
            ),
        ),
        Codec(
            "dmb",
            (".dmb",),
            _mmap_reader(_core.read_dmb),
            _file_sink_writer(_core.write_dmb),
            record=_core.DepthMap,
            datatype="depth_map",
            read_window=_mmap_selector_reader(_core.read_dmb_window),
            supported_features=(
                "scalar_float32",
                "little_endian",
                "zero_invalid",
                "pixel_windows",
            ),
            unsupported_features=(
                "normal_maps",
                "confidence",
                "embedded_scale",
            ),
        ),
    )
